src/compiler/crystal/semantic/main_visitor.cr

```diff
@@ -2010,8 +2010,11 @@ module Crystal
     def visit_allocate(node)
       instance_type = scope.instance_type
 
-      if instance_type.is_a?(GenericClassType)
+      case instance_type
+      when GenericClassType
         node.raise "can't create instance of generic class #{instance_type} without specifying its type vars"
+      when UnionType
+        node.raise "can't create instance of a union type"
```

Comment:
Well, we can define `Union.new(value : T)` in the standard library. This just checks that if we end up calling `allocate` it gives that error.